test/unit/math/rev/mat/fun/gp_matern52_cov_test.cpp

```diff
@@ -0,0 +1,36 @@
+#include <cmath>
+#include <gtest/gtest.h>
+#include <limits>
+#include <stan/math/rev/mat.hpp>
+#include <string>
+#include <vector>
+
+TEST(MathPrimMat, vec_double_gp_matern52_cov1) {
+  stan::math::var sigma = 0.2;
+  stan::math::var l = 5.0;
+
+  std::vector<stan::math::var> x(3);
+  x[0] = -2;
+  x[1] = -1;
+  x[2] = -0.5;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, Eigen::Dynamic> cov;
+  EXPECT_NO_THROW(cov = stan::math::gp_matern52_cov(x, sigma, l));
```

Comment:
The way to check that return types are what you expect is:

```
auto y = 1.0 / 1; // no need to try to wrap this part in EXPECT_NO_THROW -- that's already tested elsewhere
EXPECT_TRUE((std::is_same<double, decltype(y)>::value));
```

Not necessary to check values in this instance, but if you want to it's fine. If x, sigma, and l are all var types, the thing to test is that:

```
gp_matern52_cov(value_of(x), value_of(sigma), value_of(l))
gp_matern52_cov(x, value_of(sigma), value_of(l))
gp_matern52_cov(value_of(x), sigma, value_of(l))
gp_matern52_cov(value_of(x), value_of(sigma), l)
```

Return the right types (first should have double scalar types, the last three should have vars).

Reply:
yeah I can't get any of these assertions to compile when comparing with the datatype of what's supposed to be a `stan::math::var`...  this works fine: `EXPECT_TRUE((std::is_same<double, decltype(cov1(i, j).val())>::value));`, but when I try to do,`EXPECT_TRUE((std::is_same<stan::math::var, decltype(cov2(i, j))>::value));`  this either doens't compile or fails every time. I have to come bacl to it. Are you sure `is_same` works for datatpes that are not standard datatypes or should I use boost assertions is true?

Reply:
I tried using `boost::is_same` assertion, is doesn't work. Has anyone found a type assertion that works for our `stan::math::vars`?

Reply:
Oh weird. I'll look at this later.

Reply:
YoyoyoyoyoYOYOOoyo, this should be a reference I think! So is_same is sensitive to this stuff.

```EXPECT_TRUE((std::is_same<stan::math::var&, decltype(cov2(i, j))>::value));```

```operator()(int i, int j)``` is going to return a writable reference! (https://eigen.tuxfamily.org/dox/classEigen_1_1DenseCoeffsBase_3_01Derived_00_01WriteAccessors_01_4.html)

> my intent was not for you to solve this for me

Nah nah, team effort yo. It's all good.

Reply:
alright so here's the summary: all of the declarations above pass for `is_same<stan::math::var&, dec...` but the first one should pass for `double`.... so in general I'm always outputting references to `vars`. However, if I do `cov(i, j).val()` this passes for is same double? does this sound about right?  

how would you like me to proceed?

Reply:
The goal of these checks is just to verify that when any of the inputs have vars, then the outputs are vars (just to check the type promotion that you recently fixed works). Since you're checking types, that's should just be a few lines of code to make the output, and then a line to make sure it's the right thing. If you've got checks in place that cover that, then groovy.

To check outputs are vars, it makes more sense to check that cov(i, j) are var references than cov(i, j).val() are doubles.